tgui/packages/tgui/interfaces/ShuttleBlueprints.tsx

```diff
@@ -0,0 +1,458 @@
+  [Direction.NORTH, 'up'],
+  [Direction.SOUTH, 'down'],
+  [Direction.EAST, 'right'],
+  [Direction.WEST, 'left'],
+];
+
+const DirectionPad = (props: DirectionPadProps) => {
+  const { title, tooltip, enabledDirections, selectedDirection, onSelect } =
+    props;
+  const [north, south, east, west] = directionData.map(
+    ([direction, icon_suffix], i) => (
+      <Stack.Item key={i}>
+        <Button
+          fluid
+          m={0}
+          icon={`arrow-${icon_suffix}`}
+          selected={selectedDirection & direction}
+          disabled={!(enabledDirections & direction)}
+          onClick={() => onSelect(direction)}
+        />
+      </Stack.Item>
+    ),
+  );
+  const titleNode = (
+    <Box width="100%" textAlign="center">
+      {title}
+    </Box>
+  );
+  return (
+    <Section
+      fill
+      title={
+        tooltip ? <Tooltip content={tooltip}>{titleNode}</Tooltip> : titleNode
+      }
+    >
+      <Stack fill vertical align="center" justify="center">
+        {north}
+        <Stack.Item>
+          <Stack>
+            {west}
+            <Stack.Item width="1rem" mx={1} />
+            {east}
+          </Stack>
+        </Stack.Item>
+        {south}
+      </Stack>
+    </Section>
+  );
+};
+
+const VisualizationToggle = (props: VisualizationToggleProps) => {
+  const { visualizing } = props;
+  const { act } = useBackend<ShuttleBlueprintsData>();
+  return (
+    <Tooltip
+      content="Toggle a visualization of shuttle frames you can use to construct a shuttle.
+                Red tiles indicate frame parts built in invalid areas,
+                or parts of suitable areas that need frame parts built on them."
+    >
+      <Box inline>
+        Visualization:
+        <Button
+          color="transparent"
+          icon={visualizing ? 'toggle-on' : 'toggle-off'}
+          onClick={() => act('toggleVisualization')}
+        />
+      </Box>
+    </Tooltip>
+  );
+};
+
+const ShuttleConstruction = (props: ShuttleConstructionProps) => {
+  const [shuttleDirection, setShuttleDirection] = useState<Direction>(
+    Direction.NORTH,
+  );
+  const { act } = useBackend<ShuttleBlueprintsData>();
+  const {
+    onShuttleFrame,
+    visualizing,
+    tooManyShuttles,
+    onCustomShuttle,
+    masterExists,
+  } = props;
+  return (
+    <Stack justify="space-around">
+      <Stack.Item grow>
+        <DirectionPad
+          title="Shuttle Direction"
+          tooltip="This specifies the direction that the shuttle being built is facing."
+          enabledDirections={Direction.ALL}
+          selectedDirection={shuttleDirection}
+          onSelect={(dir) => setShuttleDirection(dir)}
+        />
+      </Stack.Item>
+      <Stack.Item>
+        <Stack fill vertical align="end" justify="space-between">
+          <Stack.Item>
+            <VisualizationToggle visualizing={visualizing} />
+          </Stack.Item>
+          <Stack.Item>
+            <Stack vertical>
+              <Stack.Item>
```

Comment:
this isn't a suggestion, but i wish for a world where we didn't have to spam propless `Stack.Item`s simply because I think the nesting sucks and it's probably one of the reasons people say that tgui coding is bad

However I'm pretty sure that if I suggested that, then I would get yelled at by a tgui maintainer, so I wouldn't try to fix this

Reply:
One day we will just have the gap property